### PR TITLE
P6 grid

### PR DESCRIPTION
**Pull request by vuongc**

### CODE REVIEW

File: src/components/molecules/p6-grid/locales/p6-grid.i18n.fr.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "hideColumn": "Colonnes cachées (cliquer pour les afficher)",
+  "emptyGrid": "Aucun résultats trouvés",
```
Comment (jswale) [inline src/components/molecules/p6-grid/locales/p6-grid.i18n.fr.json:3]: Singular

File: src/components/molecules/p6-grid/components/p6-grid-cell/p6-grid-cell.tsx [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+  /**
+   * Header id
+   */
+  @Prop() headerId!: string;
+
+  /**
+   * Cell label
+   */
+  @Prop() label = "-";
```
Comment (jswale) [inline src/components/molecules/p6-grid/components/p6-grid-cell/p6-grid-cell.tsx:89]: Use EMPTY_LABEL constant instead

File: src/components/molecules/p6-grid/utils.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  id: string,
+  cells: Array<HeaderCell>
+): HeaderCell {
+  const foundCell = cells.find((cell) => cell.id === id);
+
+  if (typeof foundCell === "undefined") {
+    throw new Error(
+      `getHeaderCellById: the header with id ${id} desn't exist!`
+    );
+  } else {
```
Comment (jswale) [inline src/components/molecules/p6-grid/utils.ts:53]: remove the else and do the return

File: src/components/molecules/p6-grid/utils.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  headerId: string,
+  cells: Array<RowCell>
+): RowCell {
+  const foundCell = cells.find((cell) => cell.headerId === headerId);
+
+  if (typeof foundCell === "undefined") {
+    throw new Error(
+      `getRowCellById: the cell with header id ${headerId} doesn't exist!`
+    );
+  } else {
```
Comment (jswale) [inline src/components/molecules/p6-grid/utils.ts:68]: ditto

File: src/components/molecules/p6-grid/locales/p6-grid.i18n.fr.json (hunk 1 above)
Comment (vuongc) [reply src/components/molecules/p6-grid/locales/p6-grid.i18n.fr.json:3]: done

File: src/components/molecules/p6-grid/components/p6-grid-cell/p6-grid-cell.tsx (hunk 2 above)
Comment (vuongc) [reply src/components/molecules/p6-grid/components/p6-grid-cell/p6-grid-cell.tsx:89]: done

File: src/components/molecules/p6-grid/utils.ts (hunk 3 above)
Comment (vuongc) [reply src/components/molecules/p6-grid/utils.ts:53]: done

File: src/components/molecules/p6-grid/utils.ts (hunk 4 above)
Comment (vuongc) [reply src/components/molecules/p6-grid/utils.ts:68]: done

File: src/components/molecules/p6-grid/components/p6-grid-row/p6-grid-row.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+   * Is row selected or not
+   */
+  @Prop() selected = false;
+
+  componentDidLoad(): void {
+    this.host.addEventListener("contextmenu", (event: MouseEvent) => {
+      event.preventDefault();
+      this.moveContextMenu(event.clientX, event.clientY);
+      this.contextMenuCallback();
+    });
```
Comment (vincentv) [inline src/components/molecules/p6-grid/components/p6-grid-row/p6-grid-row.tsx:31]: the `contextmenu` event Bubbles, we should put the listener on the grid and retrieve the context via the cell specified in the event.

File: src/components/molecules/p6-grid/locales/p6-grid.i18n.en.json [hunk 6]
```diff
@@ -0,0 +1,5 @@
+{
+  "hideColumn": "Hidden columns (click to re-display)",
```
Comment (vincentv) [inline src/components/molecules/p6-grid/locales/p6-grid.i18n.en.json:2]:
```suggestion
  "hideColumn": "Hidden columns (click to display)",
```

File: src/components/molecules/p6-grid/utils.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  );
+}
+
+export function getCellLabelByHeaderId(
+  cellId: string,
+  headerId: string,
+  label: string
+): string {
+  const calculatedLabel: string = isEmpty(label) ? EMPTY_LABEL : label;
+  return cellId === headerId ? calculatedLabel : EMPTY_LABEL;
```
Comment (vincentv) [inline src/components/molecules/p6-grid/utils.ts:29]:
```suggestion
  return cellId !== headerId ||  isEmpty(label) ? EMPTY_LABEL : label;
```

File: src/components/molecules/p6-grid/utils.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+export function getRowCellById(
+  headerId: string,
+  cells: Array<RowCell>
+): RowCell {
+  const foundCell = cells.find((cell) => cell.headerId === headerId);
+
+  if (typeof foundCell === "undefined") {
+    throw new Error(
+      `getRowCellById: the cell with header id ${headerId} doesn't exist!`
+    );
```
Comment (vincentv) [inline src/components/molecules/p6-grid/utils.ts:62]:
```suggestion
    throwError(
      `getRowCellById: the cell with header id ${headerId} doesn't exist!`
    );
```

File: src/utils/attribute.ts [hunk 9]
```diff
@@ -19,6 +19,10 @@ export function cleanupValue<T>(value: T | undefined | null): T | undefined {
   return isEmpty(value) ? undefined : (value as T);
 }
 
+export function isNumber(value: unknown): boolean {
```
Comment (vincentv) [inline src/utils/attribute.ts:22]:
```suggestion
export function isNumber(value: unknown): value is number {
```

File: src/components/organisms/p6-grid/components/p6-grid-header-column.tsx [hunk 10]
```diff
@@ -0,0 +1,20 @@
+import { h } from "@stencil/core";
+import { Column, DataItem } from "../core/entities";
+
+export function P6GridHeaderColumns(props: {
```
Comment (vuongc) [inline src/components/organisms/p6-grid/components/p6-grid-header-column.tsx:4]: File and class have different name

File: src/components/organisms/p6-grid/components/p6-grid-options/p6-grid-options.tsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        role="button"
+        size="small"
+        tabIndex={index}
+      >
+        {column.label}
+      </p6-tag>
+    );
+  };
+
+  private unhideColumn(column: Column<DataItem>): (event: MouseEvent) => void {
```
Comment (vuongc) [inline src/components/organisms/p6-grid/components/p6-grid-options/p6-grid-options.tsx:59]: Could used showColumn too

File: src/components/organisms/p6-grid/components/p6-grid-textarea-cell-editor/p6-grid-textarea-cell-editor.tsx [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+      row: this.row,
+      status: EditingCellStatus.Start,
+    });
+  }
+
+  render(): JSX.Element {
+    const hasError =
+      this.errorMessage !== undefined && this.errorMessage?.trim() !== "";
+    const classes = {
+      "has-tooltip-arrow": hasError,
```
Comment (vuongc) [inline src/components/organisms/p6-grid/components/p6-grid-textarea-cell-editor/p6-grid-textarea-cell-editor.tsx:69]: `"has-tooltip-arrow has-tooltip-multiline": hasError` is better

File: src/components/organisms/p6-grid/components/p6-grid-textarea-cell-editor/p6-grid-textarea-cell-editor.tsx [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    const input = event.target as HTMLTextAreaElement;
+
+    const isValid = input.checkValidity();
+    if (!isValid) {
+      this.errorMessage = input.validationMessage;
+    }
+  };
+
+  private keyUpHandler = (event: KeyboardEvent): void => {
+    const isEnterKey = event.keyCode === 13;
```
Comment (vuongc) [inline src/components/organisms/p6-grid/components/p6-grid-textarea-cell-editor/p6-grid-textarea-cell-editor.tsx:98]: You should create a constant for the Enter key and the Escape key

File: src/components/organisms/p6-grid/components/p6-grid-header-menu/p6-grid-header-menu.tsx [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+    };
+  }
+
+  private showColumnHandler(visible: boolean): (event: Event) => void {
+    return () => {
+      this.p6ShowColumn.emit({ column: { ...this.column, hidden: !visible } });
+    };
+  }
+
+  private isAlignmentDisabled(align: Alignment): boolean {
```
Comment (jswale) [inline src/components/organisms/p6-grid/components/p6-grid-header-menu/p6-grid-header-menu.tsx:166]: function naming

File: src/components/organisms/p6-grid/p6-grid.tsx [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+import {
+  compareRow,
+  filterBy,
+  fromData,
+  rangeSelectRow,
+  replaceRow,
+} from "./core/row";
+
+/**
+ * TODO :
```
Comment (jswale) [inline src/components/organisms/p6-grid/p6-grid.tsx:44]: TODO :D

**Comment by vuongc** [comment]
I Approve

**Pull request merged**